### PR TITLE
VACMS 2704 QA Standalone Page Field

### PR DESCRIPTION
**Pull request by kaise-lafrai**
## Description

See #2704 

## QA steps

As an editor
- Login

### No standalone page
- Navigate to QA node add page here : `/node/add/q_a`
- [x] Confirm you see a new boolean (checkbox) field called **Enable standalone page** 
- Do NOT check **Enable standalone page** and fill out the fields within the **Question & Answer** group at the very top
- Fill out the fields within the **Question & Answer** group at the very top
- Select an owner from the sidebar within Governance.
- [x] Select SAVE and confirm the node saves.
- [x] Scroll down on preview and confirm the **Enable standalone page** exists and underneath it says No standalone page

### Standalone page
- Navigate to QA node add page here : `/node/add/q_a`
- [x] Confirm you see a new boolean (checkbox) field called **Enable standalone page** 
- [x] Check **Enable standalone page** and confirm the Q&A Page Components fields become visible
- Select an owner from the sidebar within Governance.
- Select SAVE
- [x] Confirm that you cannot save without filling out CTA Button label + link
- Enter CTA Button label + link values
- [x] Confirm 2 errors have been found: Primary category, Topics
- Select Primary category and select a tag
- [x] Select SAVE and confirm the node saves.
- [x] Scroll down on preview and confirm the **Enable standalone page** exists and underneath it says Standalone page.

### Edit QA Form
- Navigate to the edit form from the previous steps above.
- Toggle the **Enable standalone page** checkbox. 
- [x] If checking the checkbox, confirm all required fields are presented to the user when clicking save.
- [x] If unchecking the checkbox, confirm all required fields are not required to be completed and user can save leaving the required fields empty.

### Create checklist
- Navigate to Checklist node add page here : `/node/add/checklist`
- Complete form and leave the CTA Button fields and Field Tags section at the bottom of the form empty.
- [x] Select SAVE and confirm validation message appears requiring the fields to be completed.

## Definition of Done
- [ ] Product release are written (or in progress), if required.
- [ ] Documentation has been updated, if applicable.

### CODE REVIEW

**Review by oksana-c — CHANGES_REQUESTED** [review]
@kaise-lafrai , the ACs are implemented in a bit of a reversed fashion here.

The logic should be: 
If the checkbox is checked, then the fieldset with standalone page fields is displayed.
If it's not checked, then the fieldset with standalone page fields is not displayed.

I know @acrollet offered to get this through the finish line, so tagging him here for visibility.

**Comment by acrollet** [comment]
Looking at this now...

**Comment by acrollet** [comment]
@oksana-c tests are passing and I've addressed the requirements and completed all QA steps.

**Comment by oksana-c** [comment]
@acrollet, one more update needs to happen before this goes out - CTA buttons field is required, but doesn't throw an error on validation if button paragraph fields are empty.

The requirement for buttons in all R&S nodes is - at least one CTA button should be added per node, max - two buttons.

**Comment by acrollet** [comment]
@oksana-c I assume this change should be made for all content types where field_buttons appears?

**Comment by oksana-c** [comment]
@acrollet good point 👍 . Paragraph "Button" should have both fields Label and Link marked required. Then we'd need to add logic to prevent form errors for CTA buttons on Q&A content type if the "Enable standalone page" checkbox is not checked.

**Comment by acrollet** [comment]
@oksana-c sounds good, thanks!

**Comment by acrollet** [comment]
Pushed, I'll re-validate on the CI environment after the tests run.

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 1]
```diff
@@ -297,6 +329,33 @@ function _va_gov_backend_audience_topics_validation(array $form, FormStateInterf
...
+
+    foreach ($form_state->getValue('field_buttons') as $idx => $val) {
+      if (empty($form_state->getValue('field_buttons')[$idx]['subform']['field_button_label'][0]['value'])) {
+        $form_state->setErrorByName("field_buttons][{$idx}][subform][field_button_label", t('Button Label field is required.'));
+      }
+
+      if (empty($form_state->getValue('field_buttons')[$idx]['subform']['field_button_link'][0]['uri'])) {
+        $form_state->setErrorByName("field_buttons][{$idx}][subform][field_button_link", t('Button Link field is required.'));
+      }
+    }
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:357]: if we remove this part, then we get the behavior, we're looking for.

**Review by oksana-c — CHANGES_REQUESTED** [review]

**Comment by oksana-c** [comment]
glitch in the last action taken. nm

**Comment by acrollet** [comment]
@oksana-c I think you were right... Re-validating locally.

**Comment by acrollet** [comment]
Local validation complete, will re-check on CI when ready.

**Comment by acrollet** [comment]
Re-validated on http://pr3170.ci.cms.va.gov/

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 2]
```diff
@@ -173,6 +173,24 @@ function _va_gov_backend_add_vamc_regions_select(array &$form) {
  * Implements hook_form_alter().
  */
 function va_gov_backend_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  // Add states + custom validation if the Standalone Page field is present.
+  if (isset($form['field_standalone_page'])) {
+    $form['field_primary_category']['widget']['#states']['required']['input[id="edit-field-standalone-page-value"]'] = ['checked' => TRUE];
+    $form['field_tags']['widget'][0]['subform']['field_topics']['#states']['required']['input[id="edit-field-standalone-page-value"]'] = ['checked' => TRUE];
+    $form['field_buttons']['widget'][0]['subform']['field_button_label']['widget'][0]['value']['#states']['required']['input[id="edit-field-standalone-page-value"]'] = ['checked' => TRUE];
+    $form['field_buttons']['widget'][0]['subform']['field_button_link']['widget'][0]['uri']['#states']['required']['input[id="edit-field-standalone-page-value"]'] = ['checked' => TRUE];
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:181]: `['widget'][0]` works only on the first instance of the paragraph. 
In the scenario where user removes the paragraph via "Remove" button and then adds another instance of the Button paragraph, the button fields are no longer required.

also, when the user removes the CTA paragraph and leaves the CTA field empty - a couple of notices are thrown on node save. see screenshot below

Screen Shot 2020-10-15 at 11 54 12 AM

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module [hunk 3]
```diff
@@ -173,6 +173,24 @@ function _va_gov_backend_add_vamc_regions_select(array &$form) {
...
+
+    // Add custom validation.
+    array_unshift($form['#validate'], '_va_gov_backend_standalone_page_validation');
+  }
+  else {
+    // Mark CTA button fields as required.
+    if (!empty($form['field_buttons'])) {
+      $form['field_buttons']['widget'][0]['subform']['field_button_label']['widget'][0]['value']['#required'] = TRUE;
+      $form['field_buttons']['widget'][0]['subform']['field_button_link']['widget'][0]['uri']['#required'] = TRUE;
```
Comment (oksana-c) [inline docroot/modules/custom/va_gov_backend/va_gov_backend.module:190]: same here - `['widget'][0]`  - index 0 only covers the first instance of a paragraph

File: docroot/modules/custom/va_gov_backend/va_gov_backend.module (hunk 2 above)
Comment (acrollet) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module]: @oksana-c after much hassling with #states + paragraphs + AJAX I've added a core patch which resolves these issues for me locally. (Along with improving the original logic)

Comment (acrollet) [reply docroot/modules/custom/va_gov_backend/va_gov_backend.module]: I will re-validate on the CI environment after it's been rebuilt.

**Comment by acrollet** [comment]
The CI environment behavior is not matching my local, investigating...

**Comment by acrollet** [comment]
Belay that, drupal + browser cache clears fixed it.

**Comment by acrollet** [comment]
@oksana-c I've successfully re-validated all QA steps on the CI environment. Please ensure you clear browser cache before testing removing/adding CTA buttons :-)

Also, here's a link to the issue for the new core patch: https://www.drupal.org/node/1091852

**Review by oksana-c — APPROVED** [review]
looks great 👍 thank you @acrollet !

**Pull request merged**